platform/consul/conversion.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consul
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/api"
+	"istio.io/pilot/model"
+)
+
+const (
+	protocolTagName = "protocol"
+	externalTagName = "external"
+)
+
+func convertTags(tags []string) model.Tags {
+	out := make(model.Tags, len(tags))
+	for _, tag := range tags {
+		vals := strings.Split(tag, "=")
+		if len(vals) > 1 {
+			out[vals[0]] = vals[1]
+		} else {
+			// TODO safe to assume all tags are of form "key=value"?
+			out[tag] = ""
+		}
+	}
+	return out
+}
+
+func convertPort(port int, name string) *model.Port {
+
+	// TODO default HTTP/TCP?
+	defaultName := "tcp"
+	if name == "" {
+		name = defaultName
+	}
+
+	return &model.Port{
+		Name:     name,
+		Port:     port,
+		Protocol: convertProtocol(name),
+	}
+}
+
+func convertService(endpoints []*api.CatalogService) *model.Service {
+	name, addr, external := "", "", ""
+
+	node, datacenter := "", ""
+
+	ports := model.PortList{}
+	for _, endpoint := range endpoints {
+		name = endpoint.ServiceName
+		ports = append(ports, convertPort(endpoint.ServicePort, endpoint.NodeMeta[protocolTagName]))
+
+		// TODO This will not work if service is a mix of external and local services
+		// or if a service has more than one external name
+		if endpoint.NodeMeta[externalTagName] != "" {
```

Comment:
It looks like external services in Consul are designated by using a hostname instead of an IP address (see https://www.consul.io/api/catalog.html#register-entity and https://www.consul.io/docs/guides/external.html). If that is the case, we should use ip/hostname to decide whether a service is external instead of extra meta data.